### ISSUE TITLE
Best way to paralellize Circuitscape?

### ISSUE THREAD

**Issue by VLucet** [Circuitscape/Circuitscape.jl#236]
I was wondering what would be the best way to paralellize Circuitscape, when the goal is not to solve multiple pairs per map (this is natively paralellized if I'm not mistaken), but when the goal is to solve many maps (one or two pairs per map). In my use case I have thousands of different INI files to map over.

**Comment by ViralBShah** [comment]
Use Julia's parallelization with `pmap`.

**Comment by vlandau** [comment]
A @distributed or @threads for loop might also be a viable option, though not sure if @threads would be optimal for this.

**Comment by VLucet** [comment]
@ViralBShah @vlandau I appreciate the quick reply. I will make an attempt and reopen if I run into issues.

**Comment by ViralBShah** [comment]
I'll be curious to know how both options fair. Remember to set number of blas threads to 1.

**Comment by vlandau** [comment]
> Remember to set number of blas threads to 1.

Good call. I do the same thing [in Omniscape](https://github.com/Circuitscape/Omniscape.jl/blob/master/src/run_omniscape.jl#L60), and it helps with performance.

**Comment by VLucet** [comment]
This might be a little over my head. Could you explain further what that means and how to do it?

**Comment by vlandau** [comment]
> how to do it?

BLAS is the basic linear algebra solver, and it does linear algebra in parallel using multithreading. If you are doing higher level parallelization, where each job is itself using BLAS, then you can get "oversubscription," which is basically having more work queued up to do than you have hardware resources. This causes a competition for resources that can have consequences for performance. Setting BLAS number of threads to 1 prevents this from happening. The link in my comment above shows the code to do that.

**Comment by ViralBShah** [comment]
Ideally we would have it go into the documentation - but that needs #242. Until then this will have to suffice.

**Comment by VLucet** [comment]
the following script launch with `julia -p 2`  now works: 
```
@everywhere using Pkg
@everywhere using Distributed
@everywhere using LinearAlgebra.BLAS

# Set BLAS threads to 1 to avoid oversubscription
@everywhere BLAS.set_num_threads(1)

Pkg.activate(".")
@everywhere using Circuitscape

# Read the ini files
@everywhere searchdir(path,key) = filter(x->occursin(key,x), readdir(path))
@everywhere dir = "config/ini_circuitscape/all/"
@everywhere ext = "ini"
@everywhere ini_list = dir .* searchdir(dir, ext)

pmap(compute, ini_list)
```
Im running into issues trying to get rid of the @everywhere and using `addprocs()`. each worker completes one call to `compute` and I run in the following error. Happy to post another issue or reopen this. 
```
ERROR: LoadError: On worker 3:
KeyError: key Circuitscape [2b7a1792-8151-5239-925d-e2b8fdfa3201] not found
getindex at ./dict.jl:477 [inlined]
root_module at ./loading.jl:962 [inlined]
deserialize_module at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:894
handle_deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:799
deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:722
deserialize_datatype at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:1192
handle_deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:775
deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:722
handle_deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:782
deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:722 [inlined]
deserialize_msg at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/messages.jl:99
#invokelatest#1 at ./essentials.jl:709 [inlined]
invokelatest at ./essentials.jl:708 [inlined]
message_handler_loop at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/process_messages.jl:185
process_tcp_streams at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/process_messages.jl:142
#101 at ./task.jl:333
Stacktrace:
 [1] (::Base.var"#732#734")(::Task) at ./asyncmap.jl:178
 [2] foreach(::Base.var"#732#734", ::Array{Any,1}) at ./abstractarray.jl:1920
 [3] maptwice(::Function, ::Channel{Any}, ::Array{Any,1}, ::Array{String,1}) at ./asyncmap.jl:178
 [4] wrap_n_exec_twice(::Channel{Any}, ::Array{Any,1}, ::Distributed.var"#208#211"{WorkerPool}, ::Function, ::Array{String,1}) at ./asyncmap.jl:154
 [5] #async_usemap#717(::Function, ::Nothing, ::typeof(Base.async_usemap), ::Distributed.var"#192#194"{Distributed.var"#192#193#195"{WorkerPool,typeof(compute)}}, ::Array{String,1}) at ./asyncmap.jl:103
 [6] (::Base.var"#kw##async_usemap")(::NamedTuple{(:ntasks, :batch_size),Tuple{Distributed.var"#208#211"{WorkerPool},Nothing}}, ::typeof(Base.async_usemap), ::Function, ::Array{String,1}) at ./none:0
 [7] #asyncmap#716 at ./asyncmap.jl:81 [inlined]
 [8] #asyncmap at ./none:0 [inlined]
 [9] #pmap#207(::Bool, ::Int64, ::Nothing, ::Array{Any,1}, ::Nothing, ::typeof(pmap), ::Function, ::WorkerPool, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:126
 [10] pmap(::Function, ::WorkerPool, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:101
 [11] #pmap#217(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(pmap), ::Function, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:156
 [12] pmap(::Function, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:156
 [13] top-level scope at /home/vlucet/Documents/Master/Thesis/land_con_monteregie/test.jl:18
 [14] include at ./boot.jl:328 [inlined]
 [15] include_relative(::Module, ::String) at ./loading.jl:1105
 [16] include(::Module, ::String) at ./Base.jl:31
 [17] exec_options(::Base.JLOptions) at ./client.jl:287
 [18] _start() at ./client.jl:460
in expression starting at /home/vlucet/Documents/Master/Thesis/land_con_monteregie/test.jl:18
```

**Comment by vlandau** [comment]
Make sure you run `addprocs(2)` after running all of the other code (i.e. _just_ before running `pmap`). Let me know if that solves it.

**Comment by VLucet** [comment]
It is what I was doing :) 
full script called with julia -p 2
```
using Pkg
using Distributed
using LinearAlgebra.BLAS
# Set BLAS threads to 1 to avoid oversubscription
BLAS.set_num_threads(1)

Pkg.activate(".")
using Circuitscape

# Read the ini files
searchdir(path,key) = filter(x->occursin(key,x), readdir(path))
dir = "config/ini_circuitscape/all/"
ext = "ini"
ini_list = dir .* searchdir(dir, ext)

addprocs(2)

pmap(compute, ini_list)
```

**Comment by VLucet** [comment]
not calling with `-p 2` causes another issue:
```
ERROR: LoadError: On worker 2:
KeyError: key Circuitscape [2b7a1792-8151-5239-925d-e2b8fdfa3201] not found
getindex at ./dict.jl:477 [inlined]
root_module at ./loading.jl:962 [inlined]
deserialize_module at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:894
handle_deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:799
deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:722
deserialize_datatype at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:1192
handle_deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:775
deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:722
handle_deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:782
deserialize at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Serialization/src/Serialization.jl:722 [inlined]
deserialize_msg at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/messages.jl:99
#invokelatest#1 at ./essentials.jl:709 [inlined]
invokelatest at ./essentials.jl:708 [inlined]
message_handler_loop at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/process_messages.jl:185
process_tcp_streams at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/process_messages.jl:142
#101 at ./task.jl:333
Stacktrace:
 [1] (::Base.var"#732#734")(::Task) at ./asyncmap.jl:178
 [2] foreach(::Base.var"#732#734", ::Array{Any,1}) at ./abstractarray.jl:1920
 [3] maptwice(::Function, ::Channel{Any}, ::Array{Any,1}, ::Array{String,1}) at ./asyncmap.jl:178
 [4] wrap_n_exec_twice(::Channel{Any}, ::Array{Any,1}, ::Distributed.var"#208#211"{WorkerPool}, ::Function, ::Array{String,1}) at ./asyncmap.jl:154
 [5] #async_usemap#717(::Function, ::Nothing, ::typeof(Base.async_usemap), ::Distributed.var"#192#194"{Distributed.var"#192#193#195"{WorkerPool,typeof(compute)}}, ::Array{String,1}) at ./asyncmap.jl:103
 [6] (::Base.var"#kw##async_usemap")(::NamedTuple{(:ntasks, :batch_size),Tuple{Distributed.var"#208#211"{WorkerPool},Nothing}}, ::typeof(Base.async_usemap), ::Function, ::Array{String,1}) at ./none:0
 [7] #asyncmap#716 at ./asyncmap.jl:81 [inlined]
 [8] #asyncmap at ./none:0 [inlined]
 [9] #pmap#207(::Bool, ::Int64, ::Nothing, ::Array{Any,1}, ::Nothing, ::typeof(pmap), ::Function, ::WorkerPool, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:126
 [10] pmap(::Function, ::WorkerPool, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:101
 [11] #pmap#217(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(pmap), ::Function, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:156
 [12] pmap(::Function, ::Array{String,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:156
 [13] top-level scope at /home/vlucet/Documents/Master/Thesis/land_con_monteregie/test.jl:18
 [14] include at ./boot.jl:328 [inlined]
 [15] include_relative(::Module, ::String) at ./loading.jl:1105
 [16] include(::Module, ::String) at ./Base.jl:31
 [17] exec_options(::Base.JLOptions) at ./client.jl:287
 [18] _start() at ./client.jl:460
in expression starting at /home/vlucet/Documents/Master/Thesis/land_con_monteregie/test.jl:18
```

**Comment by ViralBShah** [comment]
Can you paste your whole session?

**Comment by VLucet** [comment]
@ViralBShah Im realising i'm on 1.3.1, but you just recently commited for 1.4 to be minimum requirement. 

> Can you paste your whole session?

Do you mean my Manifest and Project files?

**Comment by ViralBShah** [comment]
> > Can you paste your whole session?
> 
> Do you mean my Manifest and Project files?

I meant the screendump of everything from the point you start julia. You could post it as a gist. That way we can see all the commands and also try reproduce. I suspect the fix should be straightforward.

**Comment by ViralBShah** [comment]
> @ViralBShah Im realising i'm on 1.3.1, but you just recently commited for 1.4 to be minimum requirement.

Yes, that is mainly because some of the dependent packages need Julia 1.4.

**Comment by VLucet** [comment]
Alright, here is a screen dump, with fresh install of julia 1.4.2 and dev circuitscape.
https://gist.github.com/VLucet/b472d55fe8202139daeedca33b734a6c

**Comment by ViralBShah** [comment]
The `addprocs(2)` should be done before the `using Circuitscape` line. That will ensure that the packages are loaded on all processors.

**Comment by VLucet** [comment]
@ViralBShah thank you for your continued help on this. I keep running into errors, as it looks like whenever I do `addprocs()`, I still have to reload Pkg, I have to activate and instantiate and also load Circuitscape as well as declare all variables with @everywhere.  This seems to be because, from the doc on addprocs, the different workers never synchronize.
```
help?> addprocs()
  addprocs(; kwargs...) -> List of process identifiers

  Equivalent to addprocs(Sys.CPU_THREADS; kwargs...)

  Note that workers do not run a .julia/config/startup.jl startup script, nor do they synchronize their global state (such as global variables, new method definitions, and loaded modules)
  with any of the other running processes.
```
For example:
https://gist.github.com/VLucet/509d269a86a28161b50ffe3818ebc5f1

**Comment by ViralBShah** [comment]
Doesn't this just work for you? I'm doing this on Julia 1.4.2. 

```
julia> using Distributed

julia> addprocs(2)
2-element Array{Int64,1}:
 2
 3

julia> using Circuitscape
```

**Comment by VLucet** [comment]
@ViralBShah I am at a loss to understand what is going on, despite my reading of the doc of the `Pkg`, but I think my issue is related to https://github.com/JuliaLang/julia/issues/28781. 

Maybe there is something very basic that I do not understand, and to the risk of looking stupid, here is what I did:
I run julia in a container (fresh install of 1.4.2, nothing is installed yet at the beginning of my session). I activate and instantiate from a `Manifest` that uses a specific sha of the Circuitscape repo (one of the recent commits). Here is the gist:
https://gist.github.com/VLucet/30eb4e82f342e7a0a210d7cff2b61680
What is odd is that `using Circuitscape` only worked when I installed Circuitscape in the non-project environment (default environment), even if I installed a different version in the non-project (default) environment. This seems to be because workers inherit the default environment instead of the project environment, and will only not throw an error if the right packages are also present in the default environment, regardless of their versions.

**Comment by VLucet** [comment]
And, as per suggested in the issue, I fixed my problem with `addprocs(cores, exeflags="--project")`

**Comment by ViralBShah** [comment]
@kristofferC - Is this diagnosis right about the error described above due to Pkg environments issue in https://github.com/JuliaLang/julia/issues/28781?

@VLucet Maybe you just have to throw `@everywhere` on all the things for now.

**Comment by ranjanan** [comment]
Have you guys checked out [this code](https://github.com/Circuitscape/Circuitscape.jl/compare/RA/metapar) on RA/metapar. It adds a flag to the INI which replaces the internal `pmap` call with a map. I'll make this a PR and @VLucet can test whether it works for him.

**Comment by VLucet** [comment]
@ranjanan I will give it a try! If I understand I have to add `meta_parallelize = True` to my INI files and then I have to call `compute` on an array of file names?

**Comment by aviks** [comment]
I see you are trying to remove `@everywhere` -- that is a bad idea, that invocation is required. 

The way this will work is that you start with multiple workers, `julia --machinefile=...` (or `julia -p ..` if its only one machine`, and then on the master node, in your top level file, run something like this (simplifying)

```
@everywhere activate("/path/to/project")
@everywhere using Circuitscape

data = init_data() # on master
pmap(Circuitscape.fn, data)
```

This code should probably be in a separate main or runner file, and not be part of the module. That makes things easier.

**Comment by frederikvand** [comment]
This seems to work on the cluster but its not very stable. 

import Pkg; Pkg.add("Distributed")
using Distributed

addprocs(3)
@everywhere using Pkg
@everywhere Pkg.add(PackageSpec(name="Circuitscape", rev="master"))
@everywhere using Circuitscape

cd("/scratch/leuven/330/vsc33060/circuitscape/05_current_connect_parts/06_julia/p3_pairs")

pmap(include,["01_current_p3_pairs48.jl", "01_current_p3_pairs49.jl", "01_current_p3_pairs50.jl"])

**Comment by VLucet** [comment]
@frederikvand I do also seem to observe that although all the workers start, a lot of them seem to quickly die, or seem to alternate between in use and idle (looking at your edited message suggest you had some trouble with that too). Is `batch_size` able to fix that?

**Comment by frederikvand** [comment]
Yeah I think its related to memory usage peaks. Its hard to notice with memory monitoring because the failing workers are caused by the peaks I think. When I alter batch size and use bigmemory nodes (264 gig ram) I think it is solved. They are still running so I am not sure yet!
